src/goto-harness/recursive_initialization.cpp

```diff
@@ -136,5 +151,180 @@ void recursive_initializationt::initialize_nondet(
   const recursion_sett &known_tags,
   code_blockt &body)
 {
-  body.add(code_assignt{lhs, side_effect_expr_nondett{lhs.type()}});
+  if(
+    lhs.id() != ID_symbol ||
+    !is_array_size_parameter(to_symbol_expr(lhs).get_identifier()))
+  {
+    body.add(code_assignt{lhs, side_effect_expr_nondett{lhs.type()}});
+  }
+}
+
+void recursive_initializationt::initialize_array(
+  const exprt &array,
+  std::size_t depth,
+  const recursion_sett &known_tags,
+  code_blockt &body)
+{
+  PRECONDITION(array.type().id() == ID_array);
+  const auto &array_type = to_array_type(array.type());
+  const auto array_size =
+    numeric_cast_v<std::size_t>(to_constant_expr(array_type.size()));
+  for(std::size_t index = 0; index < array_size; index++)
+  {
+    initialize(
+      index_exprt(array, from_integer(index, size_type())),
+      depth,
+      known_tags,
+      body);
+  }
+}
+
+bool recursive_initializationt::should_be_treated_as_array(
+  const irep_idt &array_name) const
+{
+  return initialization_config.pointers_to_treat_as_arrays.find(array_name) !=
+         initialization_config.pointers_to_treat_as_arrays.end();
+}
+
+bool recursive_initializationt::is_array_size_parameter(
+  const irep_idt &cmdline_arg) const
+{
+  return initialization_config.variables_that_hold_array_sizes.find(
+           cmdline_arg) !=
+         initialization_config.variables_that_hold_array_sizes.end();
+}
+
+optionalt<irep_idt> recursive_initializationt::get_associated_size_variable(
+  const irep_idt &array_name) const
+{
+  return optional_lookup(
+    initialization_config.array_name_to_associated_array_size_variable,
+    array_name);
+}
+
+void recursive_initializationt::initialize_dynamic_array(
+  const exprt &pointer,
+  std::size_t depth,
+  const recursion_sett &known_tags,
+  code_blockt &body)
```

Comment:
❓ Why have an `out` argument, when you could return the list of statements to add.

Reply:
It's a design decision. We could go that way, for sure, but in our initial design of goto-harness we went this way, and now we want to keep it for interface uniformity.

Reply:
But this is an implementation function - not a part of the interface, so you're not limited in that regard. 🤔